sql/moz-fx-data-marketing-prod/ga_derived/downloads_with_attribution_v2/query.sql

```diff
@@ -0,0 +1,367 @@
+-- Query for ga_derived.downloads_with_attribution_v2
+CREATE TEMP FUNCTION normalize_browser(browser STRING) AS (
+  CASE
+    WHEN `moz-fx-data-shared-prod.udf.ga_is_mozilla_browser`(browser)
+      THEN 'Firefox'
+    WHEN browser IN ('Internet Explorer')
+      THEN 'MSIE'
+    WHEN browser IN ('Edge')
+      THEN 'Edge'
+    WHEN browser IN ('Chrome')
+      THEN 'Chrome'
+    WHEN browser IN ('Safari')
+      THEN 'Safari'
+    WHEN browser IN ('(not set)')
+      THEN NULL
+    WHEN browser IS NULL
+      THEN NULL
+    ELSE 'Other'
+  END
+);
+
+WITH all_hits AS (
+  SELECT
+    clientId AS client_id,
+    visitId AS visit_id,
+    hit.appInfo.landingScreenName AS landing_page,
+    hit.page.pagePath AS pagepath,
+    CASE
+      WHEN (
+          hit.type = 'EVENT'
+          AND hit.eventInfo.eventAction = 'Firefox Download'
+          AND hit.eventInfo.eventCategory IS NOT NULL
+          AND hit.eventInfo.eventLabel LIKE 'Firefox for Desktop%'
+        )
+        THEN TRUE
+      ELSE FALSE
+    END AS has_ga_download_event,
+    hit.type AS hit_type,
+    CASE
+      WHEN (hit.type = 'EVENT' AND hit.eventInfo.eventAction = 'Stub Session ID')
+        THEN hit.eventInfo.eventLabel
+      ELSE NULL
+    END AS download_session_id
+  FROM
+    `moz-fx-data-marketing-prod.65789850.ga_sessions_*` AS ga
+  CROSS JOIN
+    UNNEST(hits) AS hit
+  WHERE
+    _TABLE_SUFFIX
+    BETWEEN FORMAT_DATE('%Y%m%d', DATE_SUB(@download_date, INTERVAL 2 DAY))
+    AND FORMAT_DATE('%Y%m%d', DATE_ADD(@download_date, INTERVAL 1 DAY))
+),
+pageviews AS (
+  SELECT
+    client_id AS client_id,
+    visit_id AS visit_id,
+    COUNT(pagePath) AS pageviews,
+    COUNT(DISTINCT pagePath) AS unique_pageviews,
+  FROM
+    all_hits
+  WHERE
+    hit_type = 'PAGE'
+  GROUP BY
+    client_id,
+    visit_id
+),
+hits_fields AS (
+  SELECT
+    client_id AS client_id,
+    visit_id AS visit_id,
+    mozfun.stats.mode_last(ARRAY_AGG(download_session_id)) AS download_session_id,
```

Comment:
nit: needs order by to be idempotent (these and all others)

Reply:
This should `IGNORE NULL`s, right? Since most hits don't have `download_session_id`?

Reply:
> nit: needs order by to be idempotent (these and all others)

You are referring to the `client_id` and `visit_id` at lines 63,64?

Reply:
> This should `IGNORE NULL`s, right? Since most hits don't have `download_session_id`?

If I add `download_session_id IS NOT NULL` then the hit rows indicating if a download event occurred will be removed resulting in `has_ga_download_event` being set to false.  I will add the filter `hit_type = 'EVENT' ` which will not drop the needed hit rows..

Reply:
filter `hit_type = 'EVENT' ` has been added.